### PR TITLE
Feat: simulate multiple transactions RPC call

### PR DESCRIPTION
**Pull request by stegaBOB**
#### Problem
Currently, if you want to sign multiple transactions at once, wallets will send multiple simulateTransaction calls on each transaction, and if any of the transactions are dependent on previous transactions, the simulation will fail. This results in a "Transaction may fail to complete" error message on most wallets if some of the dependent simulations fail.

#### Summary of Changes
This PR adds support for simulating multiple dependent transactions in one call. It utilizes the `account_override` field in the `load_and_execute_transactions` Bank method to update the mutable accounts after each subsequent transaction.

TODO: 
- [x] Add config values to limit max transactions to simulate
- [ ] Limit the max number of config_accounts that are passed in
- [ ] Figure out the performance overhead and potentially add some benchmarks here
- [ ] More complicated test cases

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -4451,10 +4451,117 @@ impl Bank {
...
+                }
+                TransactionExecutionResult::NotExecuted(_) => (None, None),
+            };
+            let logs = logs.unwrap_or_default();
+            post_simulation_accounts
+                .iter()
+                .enumerate()
+                .for_each(|(index, (key, account))| {
+                    if transaction.message().is_writable(index) {
+                        account_overrides.set_account(key, Some(account.clone()));
```
Comment (buffalu) [inline runtime/src/bank.rs:4539]: this will be incorrect if tx failed. fees subtracted in bank.commit_transaction

**Comment by buffalu** [comment]
another reference implementation: https://github.com/jito-foundation/jito-solana/blob/e3f294ffa1d3856048189b3e47d92df7b53aae2b/rpc/src/rpc.rs#L3990

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -4451,10 +4451,117 @@ impl Bank {
...
+            .collect();
+        let account_nums = account_keys
+            .iter()
+            .map(|keys| keys.len())
+            .collect::<Vec<_>>();
+        let mut account_overrides = self.get_account_overrides_for_simulation(&account_keys);
+
+        let mut results = vec![];
+        for (transaction, number_of_accounts) in transactions.iter().zip(account_nums) {
+            let batch = self.prepare_unlocked_batch_from_single_tx(&transaction);
```
Comment (buffalu) [inline runtime/src/bank.rs:4481]: optimization here could be locking as many non-conflicting transactions that you can + doing some batching

**Comment by stegaBOB** [comment]
Thanks @buffalu! Reviewing the Jito implementation and fixing the issues you brought up!

**Comment by stegaBOB** [comment]
What would the best way to benchmark the performance impacts an rpc change like this would have? I added some minimal rpc tests to make sure my code sorta works, but obviously that isn't sufficient for making sure this doesn't slow the validator down. Are there any existing rpc/performance benchmarks I could replicate for this? I couldn't find any in the rpc integration tests. If anyone could point me in the right direction for this that would be amazing!

**Comment by stegaBOB** [comment]
Planning on working on this again. Is there a way to unmark this PR as stale without having to reopen it again?

**Comment by willhickey** [comment]
This repository is no longer in use. Please re-open this pull request in the agave repo: https://github.com/anza-xyz/agave

**Pull request closed without merging**